website/docs/configuration/types.html.md

```diff
@@ -0,0 +1,187 @@
+    brackets containing a comma-separated series of types. Values that match the
+    tuple type must have _exactly_ the same number of elements (no more and no
+    fewer), and the value in each position must match the specified type for
+    that position.
+
+For example: an object type of `object({name=string,age=number})` would match a
+value like the following:
+
+```hcl
+{
+  name = "John"
+  age  = 52
+}
+```
+
+And a tuple type of `tuple([string, number, bool])` would match a value like the
+following:
+
+```hcl
+["a", 15, true]
+```
+
+### Complex Type Literals
+
+The Terraform language has literal expressions for creating tuple and object
+values, which are described in [Expressions: Literal
+Expressions](./expressions.html#literal-expressions) as "list/tuple" literals
+and "map/object" literals, respectively.
+
+Terraform does _not_ provide any way to directly represent lists, maps, or sets.
+However, due to the automatic conversion of complex types (described below), the
+difference between similar complex types is almost never relevant to a normal
+user, and most of the Terraform documentation conflates lists with tuples and
+maps with objects. The distinctions are only useful when restricting input
+values for a module or resource.
+
+### Conversion of Complex Types
+
+Similar kinds of complex types (list/tuple/set and map/object) can usually be
+used interchangeably within the Terraform language, and most of Terraform's
+documentation glosses over the differences between the kinds of complex type.
+This is due to two conversion behaviors:
+
+* Whenever possible, Terraform converts values between similar kinds of complex
+  types if the provided value is not the exact type requested. "Similar kinds"
+  is defined as follows:
+    * Objects and maps are similar.
+    * Tuples and lists are similar.
+    * Sets are _almost_ similar to both tuples and lists. When a value is
+      converted to a set, duplicate values are discarded and the ordering of
+      elements is lost. When a set is converted to another type, no particular
+      ordering of elements is guaranteed.
+* Whenever possible, Terraform converts _element values_ within a complex type,
+  either by converting complex-typed elements recursively or as described above
+  in [Conversion of Primitive Types](#conversion-of-primitive-types).
+
+For example: if a module argument requires a value of type `list(string)` and a
+user provides the tuple `["a", 15, true]`, Terraform will internally transform
+the value to `["a", "15", "true"]` by converting the elements to the required
+`string` element type. Later, if the module uses those elements to set different
+resource arguments that require a string, a number, and a bool (respectively),
+Terraform will automatically convert the second and third strings to the
+required types at that time, since they contain valid representations of a
+number and a bool.
```

Comment:
I think this is aligned with the example you walked me through, but I might have forgotten some subtlety.

Reply:
This is all correct.